flytepropeller/pkg/resolver/k8s_resolver.go

```diff
@@ -0,0 +1,182 @@
+		return targetInfo{}, fmt.Errorf("target %s must specify a service", &target.URL)
+	}
+
+	return targetInfo{
+		serviceName:      service,
+		serviceNamespace: namespace,
+		port:             port,
+	}, nil
+}
+
+// Build creates a new resolver for the given target, e.g. k8s:///flyteagent:flyte:8000.
+func (b *kubeBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts resolver.BuildOptions) (resolver.Resolver, error) {
+	ti, err := parseResolverTarget(target)
+	if err != nil {
+		return nil, err
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	r := &kResolver{
+		target:    ti,
+		ctx:       ctx,
+		cancel:    cancel,
+		cc:        cc,
+		k8sClient: b.k8sClient,
+	}
+	go wait.Until(r.run, 0, ctx.Done())
+	return r, nil
+}
+
+// Scheme returns the scheme supported by this resolver.
+// Scheme is defined at https://github.com/grpc/grpc/blob/master/doc/naming.md.
+func (b *kubeBuilder) Scheme() string {
+	return b.schema
+}
+
+type kResolver struct {
+	target    targetInfo
+	ctx       context.Context
+	cancel    context.CancelFunc
+	cc        resolver.ClientConn
+	k8sClient *kubernetes.Clientset
+	// wg is used to enforce Close() to return after the watcher() goroutine has finished.
+	wg sync.WaitGroup
+}
+
+// ResolveNow is a no-op at this point.
+func (k *kResolver) ResolveNow(resolver.ResolveNowOptions) {
+	logger.Infof(k.ctx, "k8s resolver: resolveNow")
+}
+
+// Close closes the resolver.
+func (k *kResolver) Close() {
+	k.cancel()
+	k.wg.Wait()
+	logger.Infof(k.ctx, "k8s resolver: closed")
+}
+
+func (k *kResolver) resolve(e *v1.Endpoints) {
+	var newAddrs []resolver.Address
+	for _, subset := range e.Subsets {
+		port := k.target.port
+
+		for _, address := range subset.Addresses {
+			newAddrs = append(newAddrs, resolver.Address{
+				Addr:       net.JoinHostPort(address.IP, port),
+				ServerName: fmt.Sprintf("%s.%s", k.target.serviceName, k.target.serviceNamespace),
+				Metadata:   nil,
+			})
+		}
+	}
+	logger.Infof(k.ctx, "k8s resolver: resolved %d addresses", len(newAddrs))
+	logger.Infof(k.ctx, "k8s resolver: resolved addresses: %v", newAddrs)
+	err := k.cc.UpdateState(resolver.State{Addresses: newAddrs, ServiceConfig: k.cc.ParseServiceConfig("{\"loadBalancingConfig\": [{\"round_robin\":{}}]}")})
+	if err != nil {
+		grpclog.Errorf("k8s resolver: failed  : %v", err)
+	}
+}
+
+func (k *kResolver) run() {
+	k.wg.Add(1)
+	defer k.wg.Done()
+	logger.Infof(k.ctx, "Starting k8s resolver for %s", k.target.serviceName)
+	watcher, err := k.k8sClient.CoreV1().Endpoints(k.target.serviceNamespace).Watch(k.ctx, metav1.ListOptions{FieldSelector: "metadata.name=" + k.target.serviceName})
+	if err != nil {
+		grpclog.Errorf("k8s resolver: failed to create watcher: %v", err)
+		return
+	}
+
+	for {
+		select {
+		case <-k.ctx.Done():
+			logger.Infof(k.ctx, "k8s resolver: context done")
+			return
+		case event, ok := <-watcher.ResultChan():
+			if !ok {
+				logger.Infof(k.ctx, "k8s resolver: watcher closed")
```

Comment:
Is the watcher reliable? In case something breaks temporarily, we would stop resolving.

Reply:
Maybe we could give it a retry/re-watch or something? I don't know whether watcher is more reliable these days, but my experience with it back in the day was not very pleasant.

Reply:
I use `wait.until`, which will recreate watcher if it's disconnected due to server timeout or something else.

Reply:
```go
go wait.Until(r.run, 0, ctx.Done())
```

Reply:
Ah, I missed that. Sorry.